SCMOD-10362: Move from nashorn to graal script engine

dgibson-microfocus opened the pull request:
https://portal.digitalsafe.net/browse/SCMOD-10362

dgibson-microfocus committed:
SCMOD-10362: Move from nashorn to graal script engine

buildmachine-sou-jenkins2 commented:
A developer build has not yet been created for this branch. Click [here](http://sou-jenkins2.swinfra.net:8080/job/Create-New-Build/parambuild/?GITHUB_SERVER=github.com&GITHUB_ORGANISATION=CAFDataProcessing&REPO_NAME=worker-document&BRANCH_NAME=SCMOD-10362) to go ahead and create the build...

buildmachine-sou-jenkins2 commented:
CI Build Link:
http://sou-jenkins2.swinfra.net:8080/job/CAFDataProcessing/job/CAFDataProcessing~worker-document~SCMOD-10362~CI

dgibson-microfocus committed:
fix

dgibson-microfocus committed:
Grant permissions to JS

dgibson-microfocus committed:
Allow IO access

dgibson-microfocus committed:
allow loading from classpath

dgibson-microfocus committed:
Directly create the graal.js engine to configure all required settings

dgibson-microfocus committed:
enable nashorn syntax extensions

dermot-hardy committed:
Put dependency management in alphabetical order

dermot-hardy committed:
Group the compile-time dependencies together

dermot-hardy committed:
Keep to the local import order policy

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/impl/ScriptImpl.java:
```diff
@@ -237,18 +238,18 @@ public void raiseEvent(final String event, final Object... args)
         }
 
         // Call the JavaScript function with the specified arguments
         // Nashorn automatically wraps checked exceptions in a RuntimeException
```
Does Graal.js behave the same w.r.t. checked exceptions?

dgibson-microfocus replied:
Yes, any exception that's thrown from a script is wrapped as a `PolyglotException` which extends `RuntimeException`

dgibson-microfocus replied:
I should also point it it wraps _every_ exception, not just checked exceptions unlike Nashorn which would throw the original `RuntimeException`

dermot-hardy replied:
Ok.  Don't forget to update the comment if necessary.

dermot-hardy committed:
Keep within the 138-character guideline

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/specs/UrlScriptSpec.java:
```diff
@@ -43,10 +44,11 @@ public UrlScriptSpec(final URL url) throws URISyntaxException
+            return compiler.compile(reader);
+        } catch (final IOException e) {
+            throw new ScriptException(e);
+        }
```
With these changes this method is now no different to the method it is overriding.  As such it should simply be removed.  Overriding it no longer makes sense.

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/specs/UrlScriptSpec.java:
```diff
@@ -77,7 +79,7 @@ public boolean isStatic()
     protected Reader openReader() throws IOException
     {
-        return new URLReader(url);
+        return new BufferedReader(new InputStreamReader(url.openStream()));
```
You should always specify a character set when creating an `InputStreamReader`.  You never want to have any platform-specific behaviour.

```suggestion
        return new BufferedReader(new InputStreamReader(url.openStream(), StandardCharsets.UTF_8));
```

dermot-hardy committed:
JavaScriptEngine: Minor whitespace change

dermot-hardy committed:
JavaScriptEngine: Minor change to add a space before the comments

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/JavaScriptEngine.java:
```diff
@@ -33,7 +35,16 @@
+                .allowHostClassLookup(s -> true) // Allow JS access to public Java classes
+                .allowIO(true) // Allow JS IO access to load additional scripts
+                .option("js.load-from-classpath", "true") // Allow JS to load files from the classpath
+                .option("js.syntax-extensions", "true"));
```
Are there more options that can be turned on so that changes to the actual scripts are not required?  I sort of half expected to see a "nashorn-compat=true" or something like that.  If there is nothing like that and changes to the scripts must be made then we'll have to treat this as a breaking change - major version number updated, break detailed in release notes, etc.  With the nature of the breaking change it's probably also a knock-on breaking change, i.e. the workers consuming the new framework version are also breaking updates since they no longer support input messages that were supported by the previous version.  We'll have to think carefully about how to roll this out (there is an argument for changing the queue names with a change like this, i.e. allowing the existing versions to consume remaining existing messages).

dermot-hardy replied:
Is this switch useful?
https://github.houston.softwaregrp.net/Verity/darwin-search-and-analytics/blob/v3.3.0/schema/pom.xml#L36

dermot-hardy replied:
Also, like I was saying if possible I'd prefer if we didn't have to change the scripts, but if that's not possible and we have to change them anyway, should we re-evaluate whether we should turn all of these on.  For example I don't know what syntax-extensions=true does but if it is something like using a Nashorn specific for loop construct instead of the JavaScript standard construct then perhaps we should leave it off and just update the scripts to be more standard.  (But like I say that's only a thought for if we have to change the scripts anyway.)

dgibson-microfocus replied:
There is a flag for backwards compatibility with Nashorn, but I assumed the point of this ticket was to start to move away from all our nashorn specific code (which is a reasonably large amount). But as you point out if that's a breaking change I'll set the compatibility flag on and that should avoid the need to change the scripts.

dgibson-microfocus replied:
@dermot-hardy  I just tried it out, even with Nashorn compatibility enabled there still needs to be script changes. The way graal.js handles numeric types and function overloading is too different from Nashorn. Though it's less than I've currently made.

Regarding `syntax-extensions` I had been updating the scripts to the correct syntax however I had to switch that on for the workflow as the way we definite and execute the `conditionFunction` of workflow actions is Nashorn specific syntax

dgibson-microfocus committed:
code review

dgibson-microfocus committed:
This is a breaking change

dgibson-microfocus committed:
Support nashorn and graal JS engines

dgibson-microfocus committed:
Execute the script using the correct engine impl

dgibson-microfocus committed:
Add engine type to script interface

dermot-hardy commented on worker-document-shared/src/main/java/com/hpe/caf/worker/document/DocumentWorkerScript.java:
```diff
@@ -26,4 +26,5 @@
     public String script;
     public String storageRef;
     public String url;
+    public String engine;
```
Update the worker-document-shared documentation for this:
[worker-document-shared#custom-document-scripts](https://github.com/CAFDataProcessing/worker-document/tree/v4.4.0/worker-document-shared#custom-document-scripts)

dermot-hardy commented on worker-document-interface/src/main/java/com/hpe/caf/worker/document/model/ScriptEngineType.java:
```diff
@@ -0,0 +1,22 @@
+ */
+package com.hpe.caf.worker.document.model;
+
+public enum ScriptEngineType
```
Interfaces, methods, parameters, etc. in the interface module should have comprehensive JavaDoc since they're used by all the workers (i.e. not just internal code).

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/specs/AbstractScriptSpec.java:
```diff
@@ -51,14 +53,21 @@ public static AbstractScriptSpec create(final ApplicationImpl application, final
+        if (script.engine == null || script.engine.isEmpty()) {
+            engineType = ScriptEngineType.NASHORN;
+        } else {
+            engineType = ScriptEngineType.valueOf(script.engine);
```
You haven't validated the engine yet so you don't know that this won't result in an `IllegalArgumentException`.  That is an unchecked exception that should only be happening when there is a fault in the code.  Although catching it here and re-throwing an `InvalidScriptException` would be fine rather than introducing earlier validation code.

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/specs/AbstractScriptSpec.java:
```diff
@@ -51,14 +53,21 @@ public static AbstractScriptSpec create(final ApplicationImpl application, final
         }
 
+        final ScriptEngineType engineType;
+        if (script.engine == null || script.engine.isEmpty()) {
```
Why make empty string valid?

```suggestion
        if (script.engine == null) {
```

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/specs/UrlScriptSpec.java:
```diff
@@ -16,37 +16,37 @@
 
-    @Nonnull
-    @Override
-    public CompiledScript compile(final Compilable compiler) throws ScriptException
```
Shouldn't this implementation still be present for the Nashorn support?

dgibson-microfocus replied:
Should it? I thought it was only overriden to avoid the close on the `URLReader` which isn't being used anymore

dermot-hardy replied:
Since we are keeping Nashorn around I'd rather that we left the existing implementation alone (because Nashorn's `URLReader` does have some Nashorn-specific enhancements).  So just rename the existing `UrlScriptSpec` class to `NashornUrlScriptSpec` and then introduce a separate new more general `UrlScriptSpec` that uses the `InputStreamReader` like you are doing.

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/specs/InlineScriptSpec.java:
```diff
@@ -25,10 +27,18 @@
     private final String script;
+    private final ScriptEngineType engineType;
 
     public InlineScriptSpec(final String script)
```
Are these constructors still required?

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/JavaScriptEngine.java:
```diff
@@ -24,24 +26,41 @@
+
+        this.graalEngineBindings = graalEngine.getBindings(ScriptContext.ENGINE_SCOPE);
+        this.nashornEngine = new ScriptEngineManager().getEngineByName("nashorn");
+        this.nashornEngineBindings = nashornEngine.createBindings();
```
Why is this changing from `.getBindings(ScriptContext.ENGINE_SCOPE);` like it was before?

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/JavaScriptEngine.java:
```diff
@@ -24,24 +26,41 @@
+    private final Bindings nashornEngineBindings;
     private final Object scriptEngineBindingsLock;
 
     public JavaScriptEngine()
```
This is always going to load both JavaScript engines.  It would be better to have separate `NashornJSEngine` and a `GraalJSEngine` classes, and only construct them if they are used.  Currently the `JavaScriptEngineLazy` class ensures that the `JavaScriptEngine` is only constructed if it is used (because it is expensive) but the way you've changed this if either engine is used both are loaded.

dgibson-microfocus replied:
Oh you're right. I had originally planned that but misread what `JavaScriptEngineLazy` was doing and thought the engine was getting constructed on startup regardless

dgibson-microfocus committed:
code review

dgibson-microfocus committed:
remove random import

dgibson-microfocus committed:
Change version back to 4.5.0-SNAPSHOT as now not a breaking change.

dermot-hardy committed:
Merge branch 'develop' into SCMOD-10362

dermot-hardy committed:
Made some Readme updates

dermot-hardy committed:
ScriptImpl: Corrected import order

dermot-hardy committed:
Removed unused constructors

dermot-hardy committed:
Script.java: Put methods in alphabetical order

dermot-hardy committed:
ScriptEngineType: Fleshed out the JavaDoc a little with some links

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/specs/InlineScriptSpec.java:
```diff
@@ -25,10 +26,12 @@
 public final class InlineScriptSpec extends AbstractScriptSpec
 {
     private final String script;
+    private final ScriptEngineType engineType;
```
You're adding this new variable to this and all the other concrete implementations of `AbstractScriptSpec`, but you haven't updated any of the `equals()` or `hashCode()` methods of these classes.  Also, why add this to each of the concrete implementations instead of further up the inheritance hierarchy to share the code?

dermot-hardy replied:
This comment hasn't been addressed:

> why add this to each of the concrete implementations instead of further up the inheritance hierarchy to share the code?

dermot-hardy replied:
I've made the change in commit 096e6f2.

dermot-hardy committed:
AbstractScriptSpec: Added a comment

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/specs/AbstractScriptSpec.java:
```diff
@@ -94,6 +107,13 @@ public static AbstractScriptSpec create(final ApplicationImpl application, final
+     *
+     * @return the JavaScript engine to use for the script represented by this specification.
+     */
+    public abstract ScriptEngineType getEngineType();
```
Mark this up as `@Nonnull` since the caller can expect it to never return `null`.  That is done for all functions that return non-primitives in this module so also review any other new ones that you've added.

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/scripting/specs/InlineScriptSpec.java:
```diff
@@ -69,6 +72,12 @@ public boolean isStatic()
+    }
+
     @Override
     protected void setScriptSpecField(final DocumentWorkerScript script)
```
None of the `setScriptSpecField()` methods have been updated.  Is the engine getting set on the output messages, from the workflow worker for example?  If so then how?  I would have expected the `setScriptSpecField()` methods to be updated to do this.

dermot-hardy replied:
Actually on reflection it might be better to do it in the `AbstractScriptSpec::toDocumentWorkerScript` method, but regardless it's still not done.

dgibson-microfocus committed:
scripts: include new fields in equals & hashcode

dgibson-microfocus committed:
Add Nashorn specific UrlScriptSpec

dgibson-microfocus committed:
Set the script engine on output

dgibson-microfocus committed:
license header

dermot-hardy committed:
NashornUrlScriptSpec code review

- Organized imports
- Made class final
- Moved getEngineType() up

dermot-hardy committed:
Moved the implementation of getEngineType() up to AbstractScriptSpec

dermot-hardy committed:
AbstractScriptSpec: Only set the engine if it is not Nashorn

This is for backwards compatibility, so that the message format doesn't unnecessarily change.

dermot-hardy committed:
ScriptImpl: Move methods to their correct positions alphabetically

dermot-hardy committed:
ScriptImpl::setScriptByUrl: Updated to use NashornUrlScriptSpec if possible

dermot-hardy committed:
Update and set the message format version

dermot-hardy committed:
InlineScriptSpec: Changed the variable name to match the others

dermot-hardy committed:
RemoteScriptSpec: Made the comment a little more generic

dermot-hardy committed:
ScriptImpl: Added some missing final specifiers

dermot-hardy committed:
ScriptImpl: Made the event handlers static

dermot-hardy committed:
JavaScriptEngine: Made variables private

They're not being used in the subclasses.

dermot-hardy committed:
JavaScriptManager: Renamed variable for consistency

dermot-hardy committed:
JavaScriptEngineLazy: Renamed variable for consistency

dermot-hardy committed:
JavaScriptEngineLazy: Set concurrencyLevel to 1

Avoid the possibility of multiple threads constructing engines at the same time

dermot-hardy committed:
JavaScriptEngineLazy: Added a missing Nonnull annotation

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/impl/ScriptImpl.java:
```diff
@@ -237,6 +263,31 @@ public void raiseEvent(final String event, final Object... args)
+
+    private static void graalHandleEvent(final Object eventHandler, final Object[] args)
+    {
+        if (!(eventHandler instanceof Function)) {
```
Is this check really required?  `Function` normally only takes one argument.  Does this just happen to work because our JS event handlers all take one argument?

dgibson-microfocus replied:
The docs aren't super clear on this, but a guest language function (or 'executable' in their terminology) will implement `Function<Object[], Object>` while and object is returned as a `Map` and primitives are just mapped directly to the Java type.

I'll add a quick unit test to assert the above either way

dermot-hardy commented on worker-document/src/main/java/com/hpe/caf/worker/document/impl/ScriptImpl.java:
```diff
@@ -237,6 +263,31 @@ public void raiseEvent(final String event, final Object... args)
+            return;
+        }
+
+        final Value jsEventHandler = Value.asValue(eventHandler);
```
Does this definitely hook back into the correct context?  I looked at the JavaDoc but I couldn't tell for sure.  Did you step through it in the debugger to make sure?

dgibson-microfocus committed:
GraalScriptEngineTest: Add JS binding test

dgibson-microfocus committed:
explicitly use the engine context

dgibson-microfocus committed:
Revert "explicitly use the engine context"

This reverts commit 59c5d70d873cc6b06c244a79c69ad14a42a140f3.

dgibson-microfocus committed:
Merge branch 'develop' of https://github.com/CAFDataProcessing/worker-document into SCMOD-10362
